### PR TITLE
ipn/ipnlocal: fix test flake when we log after a test completes

### LINKED ISSUE #3020: ipn/ipnlocal: panic: Log in goroutine after TestStateMachine has completed
Test flake at 47044f3af7cf96332cddd2c6052df3fe9364d271 ....

@apenwarr, can you look?

```
panic: Log in goroutine after TestStateMachine has completed: 
(ignored) Notify{wg={0 0 0 0 map[]}}



goroutine 156 [running]:
testing.(*common).logDepth(0xc0003e01a0, {0xc0003b5f80, 0x27}, 0x3)
	/opt/hostedtoolcache/go/1.17.1/x64/src/testing/testing.go:779 +0x4c9
testing.(*common).log(...)
	/opt/hostedtoolcache/go/1.17.1/x64/src/testing/testing.go:761
testing.(*common).Logf(0x40c5ab, {0xb22f8f, 0xc000291c00}, {0xc0000466b0, 0xc000291c00, 0xc000069cf8})
	/opt/hostedtoolcache/go/1.17.1/x64/src/testing/testing.go:807 +0x4c
tailscale.com/ipn/ipnlocal.TestStateMachine.func2({{}, {0xb21c02, 0xd}, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0002364e0, 0x0, ...})
	/home/runner/work/tailscale/tailscale/ipn/ipnlocal/state_test.go:318 +0x1e3
tailscale.com/ipn/ipnlocal.(*LocalBackend).send(0x0, {{}, {0xb21c02, 0xd}, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0002364e0, ...})
	/home/runner/work/tailscale/tailscale/ipn/ipnlocal/local.go:1207 +0x13c
tailscale.com/ipn/ipnlocal.(*LocalBackend).setWgengineStatus(0xc0004026c0, 0xc00016c2c0, {0x0, 0x0})
	/home/runner/work/tailscale/tailscale/ipn/ipnlocal/local.go:630 +0x375
tailscale.com/wgengine.(*userspaceEngine).RequestStatus(0xc0003fe000)
	/home/runner/work/tailscale/tailscale/wgengine/userspace.go:1104 +0x92
created by tailscale.com/ipn/ipnlocal.(*LocalBackend).requestEngineStatusAndWait
	/home/runner/work/tailscale/tailscale/ipn/ipnlocal/local.go:2454 +0xcf
exit status 2
FAIL	tailscale.com/ipn/ipnlocal	0.584s
```

(from https://github.com/tailscale/tailscale/pull/3019/checks?check_run_id=3831283855)

### PR DESCRIPTION
**Pull request by andrew-d**
This switches from using an atomic.Bool to a mutex for reasons that are described in the commit, and should address the flakes that we're still seeing.

Fixes #3020

Change-Id: I4e39471c0eb95886db03020ea1ccf688c7564a11
Signed-off-by: Andrew Dunham <andrew@tailscale.com>

### CODE REVIEW

File: tstest/log.go [hunk 1]
```diff
@@ -145,3 +145,50 @@ func (ml *MemLogger) String() string {
 	defer ml.Unlock()
 	return ml.Buffer.String()
 }
+
+// AsyncTestLogger returns a logger.Logf that logs to t.Logf until the test
+// finishes, at which point it no longer logs anything.
+func AsyncTestLogger(t *testing.T) logger.Logf {
```
Comment (bradfitz) [inline tstest/log.go:151]: take a `testing.TB` if possible? otherwise not a big deal.

File: tstest/log.go [hunk 2]
```diff
@@ -145,3 +145,50 @@ func (ml *MemLogger) String() string {
...
+	//                                  check atomic, testFinished = no
+	//    test finishes
+	//    run t.Cleanups
+	//    set testFinished = true
+	//                                  call t.Logf
+	//                                  panic
+	//
+	// Using a mutex ensures that all actions in goroutine 1 in the
+	// sequence above occur atomically, and thus should not panic.
+	t.Cleanup(func() {
```
Comment (bradfitz) [inline tstest/log.go:188]: cute

File: tstest/log.go [hunk 3]
```diff
@@ -145,3 +145,50 @@ func (ml *MemLogger) String() string {
 	defer ml.Unlock()
 	return ml.Buffer.String()
 }
+
+// AsyncTestLogger returns a logger.Logf that logs to t.Logf until the test
```
Comment (bradfitz) [inline tstest/log.go:149]: Is "Async" the right word here?

It's still synchronous while the test is running?

WhileTestRunningLogger? Verbose but clear.

**Review by bradfitz — APPROVED** [review]

File: tstest/log.go [hunk 4]
```diff
@@ -145,3 +145,50 @@ func (ml *MemLogger) String() string {
...
+	// the mutex and then disabling logs, we know that all existing log
+	// functions have completed, and that no future calls to the logger
+	// will log something.
+	//
+	// We can't do this with an atomic bool, since it's possible to
+	// observe the following race:
+	//
+	//    test goroutine                goroutine 1
+	//    --------------                -----------
+	//                                  check atomic, testFinished = no
```
Comment (mihaip) [inline tstest/log.go:179]: Minor, but should `testFinished` be `done`?

**Pull request merged**